torch/profiler/_utils.py

```diff
@@ -194,6 +203,86 @@ def compute_idle_time(self):
 
+    def rank_events(self, length):
+        '''
+        Filter and Rank the events based on some heuristics:
+        1) Events that are in the falling phase of the queue depth.
+        2) Events that have a high idle_time, self_time difference.
+
+        Parameters:
+            length: The number of events to return.
+        '''
+
+        # Find the interval when qd is falling to 0
+        queue_depth_list = list(reversed(self.queue_depth_list))
+        qd_values = [e.queue_depth for e in queue_depth_list]
+
+        decrease_interval = []
+        for i in range(len(qd_values) - 1):
+            if qd_values[i] <= 1:
+                # Find next zero and if the max value between them exceeds
+                # the threshold, then we have a falling interval
+                for j in range(i + 1, len(qd_values)):
+                    if qd_values[j] <= 1:
+                        peak_idx = argmax(qd_values, start=i + 1, end=j)
+                        if peak_idx is None:
+                            continue
+                        # check for threshold
+                        if qd_values[peak_idx] - qd_values[i] > 3:
+                            decrease_interval.append(
+                                Interval(queue_depth_list[peak_idx].start,
+                                         queue_depth_list[i].start))
+                            i = j
+                            break
+
+        # Filter out events that are not in the decrease interval
+        event_list = [
+            event for event in self.metrics.keys()
+            if event.intervals_overlap(decrease_interval)
+        ]
+        if event_list:
+            self_time = torch.tensor(
+                [self.metrics[event].self_time_ns for event in event_list],
+                dtype=torch.float32)
+            idle_time = torch.tensor(
+                [self.metrics[event].idle_time_ns for event in event_list],
+                dtype=torch.float32)
+
+            normalized_gain = (idle_time -
+                               torch.mean(idle_time)) / torch.std(idle_time)
+            normalized_self = (self_time -
+                               torch.mean(self_time)) / torch.std(self_time)
+            heuristic_score_list = normalized_gain + normalized_self
+
+            # Sort events by heuristic
+            event_list = [
+                event
+                for _, event in sorted(zip(heuristic_score_list, event_list),
+                                       key=lambda x: x[0],
+                                       reverse=True)
+            ]
+            event_list = event_list[:length]
+        return event_list
+
+    def get_optimizable_events(self,
+                               length: int = 1,
+                               print_enable: bool = True):
+        event_list = self.rank_events(length)
+        output = ""
+        if len(event_list) == 0:
+            output += "No events to optimize\n"
```

Comment:
Unused? Generally the way to go about this is something like:
```
output = ["Optimizable events:" if event_list else "No events to optimize"]
for event in event_list:
  output.append(...)
```
One way to think about this is: the more control flow in your function, the easier it is to mess up. If everything follows the same path (and the code works for the degenerate case) it will be more robust.